File: day4.exs
defmodule Day4 do

  def solve1(list) do
    list |> Enum.map(&parse_passport/1)
        |> Enum.map(&(Map.drop(&1, ["cid"])))
        |> Enum.map(&Enum.count/1)
        |> Enum.filter(&(&1 == 7))
        |> Enum.count
  end

  def solve2(list) do
    list |> Enum.map(&parse_passport/1)
        |> Enum.map(&(Map.drop(&1, ["cid"])))
        |> Enum.filter(&(Enum.all?(&1, fn ({key, value}) -> valid_field({key, value}) end)))
        |> Enum.filter(&(Enum.count(&1) == 7))
        |> IO.inspect
        |> Enum.count

  end

  def parse_passport([key, value | list]) do
    Map.put(parse_passport(list), key, value)
  end

  def parse_passport([]) do
    %{}
  end

  def valid_field({"byr", num}), do: String.to_integer(num) in 1920..2002
  def valid_field({"iyr", num}), do: String.to_integer(num) in 2010..2020
  def valid_field({"eyr", num}), do: String.to_integer(num) in 2020..2030
  def valid_field({"hgt", num}) do
    case String.slice(num, -2..-1) do
      "in" -> String.to_integer(String.slice(num, 0..-3)) in 59..76
      "cm" -> String.to_integer(String.slice(num, 0..-3)) in 150..193
      _ -> false
    end
  end
  def valid_field({"hcl", "#" <> color}), do: String.graphemes(color) |> Enum.all?(&(&1 in String.graphemes("0123456789") or &1 in String.graphemes("abcdef")))
  def valid_field({"ecl", num}), do: num in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
  def valid_field({"pid", num}), do: String.length(num) == 9
  def valid_field(_), do: false
end


list = IO.read(:stdio, :all)
    |> String.split("\n\n")
    |> Enum.map(&(String.split(&1, [":", " ", "\n"], trim: true)))
    |> Day4.solve2
    |> IO.inspect

File: day4.jl
x = readlines("inputs/input4")

function parse_passport_list(list) 
    i = 1
    s = []
    while i <= length(list)
        current = list[i]
        str = ""
        while current != "" && i < length(list)
            str *= " " * current
            i += 1
            current = list[i]
        end
        push!(s, parse_passport(str))
        i += 1
    end
    s
end

function parse_passport(string)
    Dict((parse_field(field) for field in split(string, ' ', keepempty=false)))
end

function parse_field(string)
    split(string, ':')
end

function part1(list, keys)
    valid(p) = count(haskey.(Ref(p), keys)) >=7
    count(valid.(list))
end

function part2(list, dict)

    valid(p) = all(( (haskey(dict, key) && dict[key](value)) for (key, value) in pairs(p)  )) && (haskey(p, "cid") ? length(keys(p)) == 8 : length(keys(p)) == 7)
    count(valid.(list))
end

validation = Dict(
                "byr"=> (v -> 1920 <= parse(Int, v) <= 2002) 
             , "iyr" => (v -> 2010 <= parse(Int, v) <= 2020)
             , "eyr" => (v -> 2020 <= parse(Int, v) <= 2030)
             , "hgt" => (v -> tryparse(Int, chop(v, tail=2)) !== nothing && (occursin("cm", v) ? 
                                                    150 <= tryparse(Int, chop(v, tail=2)) <= 193
                                                    :
                                                    59 <= tryparse(Int, chop(v, tail=2)) <= 76)),
              "hcl" => (v -> occursin(r"^#[0-9a-f]{6}$", v))
              , "ecl" => (v -> v in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]), 
              "pid" => (v -> occursin(r"^[0-9]{9}$", v))
              , "cid" => (v -> true))

pw_list = parse_passport_list(x)
@time println(part1(pw_list, keys(validation)))
@time println(part2(pw_list, validation))
